sql/scan.go

```diff
@@ -682,33 +689,75 @@ func (v *qnameVisitor) Visit(expr parser.Expr, pre bool) (parser.Visitor, parser
 	if !pre || v.err != nil {
 		return nil, expr
 	}
-	qname, ok := expr.(*parser.QualifiedName)
-	if !ok {
-		return v, expr
-	}
 
-	v.err = qname.NormalizeColumnName()
-	if v.err != nil {
-		return v, expr
-	}
-	if qname.IsStar() {
+	switch t := expr.(type) {
+	case *parser.QualifiedName:
+		qname := t
+
+		v.err = qname.NormalizeColumnName()
+		if v.err != nil {
+			return nil, expr
+		}
+		if qname.IsStar() {
+			v.err = fmt.Errorf("qualified name \"%s\" not found", qname)
+			return nil, expr
+		}
+
+		desc := v.getDesc(qname)
+		if desc != nil {
+			name := qname.Column()
+			for _, col := range v.visibleCols {
+				if !equalName(name, col.Name) {
+					continue
+				}
+				return v, v.getQVal(col)
+			}
+		}
+
 		v.err = fmt.Errorf("qualified name \"%s\" not found", qname)
 		return nil, expr
-	}
 
-	desc := v.getDesc(qname)
-	if desc != nil {
-		name := qname.Column()
-		for _, col := range v.visibleCols {
-			if !equalName(name, col.Name) {
-				continue
-			}
-			return v, v.getQVal(col)
+	case *parser.FuncExpr:
+		// Special case handling for COUNT(*) and COUNT(foo.*), expanding the star
+		// into a tuple containing the columns in the primary key of the referenced
+		// table.
+		if len(t.Name.Indirect) > 0 || !strings.EqualFold(string(t.Name.Base), "count") {
+			break
 		}
+		// The COUNT function takes a single argument. Exit out if this isn't true
+		// as this will be detected during expression evaluation.
+		if len(t.Exprs) != 1 {
+			break
+		}
+		qname, ok := t.Exprs[0].(*parser.QualifiedName)
+		if !ok {
+			break
+		}
+		v.err = qname.NormalizeColumnName()
+		if v.err != nil {
+			return nil, expr
+		}
+		if !qname.IsStar() {
+			// This will cause us to recurse into the arguments of the function which
+			// will perform normal qualified name resolution.
+			break
+		}
+		// We've got either COUNT(*) or COUNT(foo.*). Retrieve the descriptor.
+		desc := v.getDesc(qname)
+		if desc == nil {
+			break
+		}
+		// Replace the function argument with a non-NULL column. We use the first
```

Comment:
> COUNT(col) returns the count of the non-NULL values.

Ah, this is the part I had overlooked.

Still, we can pick any non-NULL column, so this could be:
```go
var col ColumnDescriptor
for _, colDesc := range desc.Columns {
  if !colDesc.Nullable {
    col = colDesc
    break
  }
}
```

And I think some mention of index selection would be valuable